File: libs/core-logic/src/lib/services/lift-device.service.ts
// libs/core-logic/src/lib/services/lift-device.service.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { CommunicationAdapter } from '../adapters/communication.adapter';

// Определим константы для команд для большей читаемости
const LIFT_COMMANDS = {
  MOVE_UP: 'CMD_UP',
  MOVE_DOWN: 'CMD_DOWN',
  STOP: 'CMD_STOP',
};

@Injectable({
  providedIn: 'root',
})
export class LiftDeviceService {
  // Сервис зависит от АБСТРАКЦИИ, а не от конкретной реализации.
  constructor(private communicationAdapter: CommunicationAdapter) {}

  public get connectionState$(): Observable<boolean> {
    return this.communicationAdapter.connectionState$;
  }

  public connect(deviceName: string): Observable<boolean> {
    return this.communicationAdapter.connect(deviceName);
  }

  public disconnect(): void {
    this.communicationAdapter.disconnect();
  }

  public moveUp(): void {
    console.log('Sending command: MOVE_UP');
    this.communicationAdapter.writeData(LIFT_COMMANDS.MOVE_UP);
  }

  public moveDown(): void {
    console.log('Sending command: MOVE_DOWN');
    this.communicationAdapter.writeData(LIFT_COMMANDS.MOVE_DOWN);
  }

  public stop(): void {
    console.log('Sending command: STOP');
    this.communicationAdapter.writeData(LIFT_COMMANDS.STOP);
  }
}

File: lib/src/core_logic/services/lift_device_service.dart
// Сервис для управления устройством-подъемником.
/// Инкапсулирует прямые вызовы к адаптеру.
 import 'package:flutter_blue_plus/flutter_blue_plus.dart';
 
 import '../../core_logic/adapters/communication_adapter.dart'; 

class LiftDeviceService {
  final CommunicationAdapter _adapter;

  LiftDeviceService(this._adapter);

  /// Поток состояния подключения, проксируется из адаптера.
  Stream<bool> get connectionState => _adapter.connectionState;
  Stream<List<ScanResult>> get scanResults => _adapter.scanResults;
  
  Future<void> startScan({Duration timeout = const Duration(seconds: 5)}) => _adapter.startScan(timeout: timeout);
  
  Future<void> stopScan() => _adapter.stopScan();

  /// Подключение к устройству.
  Future<bool> connect(BluetoothDevice device) {
    return _adapter.connect(device);
  }

  /// Обнаружение сервисов
  Future<List<BluetoothService>> discoverServices() => _adapter.discoverServices();

  /// Отключение от устройства.
  void disconnect() {
    _adapter.disconnect();
  }

  /// Команда: двигаться вверх.
  void moveUp(BluetoothCharacteristic characteristic) {
    print('SERVICE: Sending command MOVE_UP');
    _adapter.writeData(characteristic, 'mu:1');
  }

  /// Команда: двигаться вниз.
  void moveDown(BluetoothCharacteristic characteristic) {
    print('SERVICE: Sending command MOVE_DOWN');
    _adapter.writeData(characteristic, 'md:1');
  }

  /// Команда: стоп.
  void stop(BluetoothCharacteristic characteristic) {
    print('SERVICE: Sending command STOP');
    _adapter.writeData(characteristic, 'stop');
  }
  
  void dispose() {
    _adapter.dispose();
  }
}